### PR TITLE
JDK-8313616: support loading library members on AIX in os::dll_load

### PR DESCRIPTION
**Pull request by MBaesken**
AIX dlopen allows loading members of a lib; the syntax for members is libname(member-object) . However this needs an additional flag RTLD_MEMBER passed to dlopen.
Example from our OpenJDK codebase : dlopen("/usr/lib/libperfstat.a(shr_64.o)", RTLD_MEMBER | RTLD_NOW);

However, currently we do not support this in os::dll_load, so member loading would fail because of the missing RTLD_MEMBER flag.
See https://www.ibm.com/docs/en/aix/7.1?topic=d-dlopen-subroutine part about RTLD_MEMBER

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313616](https://bugs.openjdk.org/browse/JDK-8313616): support loading library members on AIX in os::dll_load (**Enhancement** - P4)

### Reviewers
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)
 * @JoKern65 (no known openjdk.org user name / role)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15204/head:pull/15204` \
`$ git checkout pull/15204`

Update a local copy of the PR: \
`$ git checkout pull/15204` \
`$ git pull https://git.openjdk.org/jdk.git pull/15204/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15204`

View PR using the GUI difftool: \
`$ git pr show -t 15204`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15204.diff">https://git.openjdk.org/jdk/pull/15204.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15204#issuecomment-1670892247)

### CODE REVIEW

File: src/hotspot/os/aix/os_aix.cpp [hunk 1]
```diff
@@ -1122,8 +1120,17 @@ void *os::dll_load(const char *filename, char *ebuf, int ebuflen) {
...
-  // RTLD_LAZY is currently not implemented. The dl is loaded immediately with all its dependants.
-  void * result= ::dlopen(filename, RTLD_LAZY);
+  // RTLD_LAZY has currently the same behavior as RTLD_NOW
+  // The dl is loaded immediately with all its dependants.
+  int dflags = RTLD_LAZY;
+  // check for filename ending with ')', it indicates we want to load
+  // a MEMBER module that is a member of an archive.
+  int flen = strlen(filename);
+  if (flen > 0 && filename[flen - 1] == ')') {
+    dflags = RTLD_LAZY | RTLD_MEMBER;
```
Comment (TheRealMDoerr) [inline src/hotspot/os/aix/os_aix.cpp:1130]: Better: `dflags |= RTLD_MEMBER;`

File: src/hotspot/os/aix/os_aix.cpp [hunk 2]
```diff
@@ -1122,8 +1120,17 @@ void *os::dll_load(const char *filename, char *ebuf, int ebuflen) {
   event.set_name(filename);
 #endif
 
-  // RTLD_LAZY is currently not implemented. The dl is loaded immediately with all its dependants.
-  void * result= ::dlopen(filename, RTLD_LAZY);
+  // RTLD_LAZY has currently the same behavior as RTLD_NOW
+  // The dl is loaded immediately with all its dependants.
+  int dflags = RTLD_LAZY;
```
Comment (TheRealMDoerr) [inline src/hotspot/os/aix/os_aix.cpp:1125]: Looks ok. Maybe it would be safer to use `RTLD_NOW` in case the implementation changes at some point of time.

Comment (TheRealMDoerr) [reply src/hotspot/os/aix/os_aix.cpp:1125]: Ok.

File: src/hotspot/os/aix/os_aix.cpp (hunk 1 above)
Comment (MBaesken) [reply src/hotspot/os/aix/os_aix.cpp]: Hi Martin, I adjusted this coding.

**Comment by MBaesken** [comment]
Hi Martin and Joachim, thanks for the reviews !

/integrate

**Pull request closed without merging**